### ISSUE TITLE
[BUG] Upgrading Kubernetes Version does not retain RKE User Addons

### ISSUE THREAD

**Issue by jakefhyde** [rancher/rancher#38703]
**Rancher Server Setup**
- Rancher version: v2.6.2

**Information about the Cluster**
- Kubernetes version: v1.20.11
- Cluster Type (Local/Downstream): Downstream
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): RKE

**Describe the bug**

When upgrading Kubernetes Version for Rancher provisioned clusters, the RKE User Addons job is not replaced, which means that addons will not be retained if overwritten by RKE's other jobs.

**To Reproduce**

- Provision an RKE1 cluster
- Add the following under `rancher_kubernetes_engine_config`:
```
ing under rancher_kubernetes_engine_config:
addons: |-
    ---
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: coredns
      namespace: kube-system
    data:
      Corefile: |
        .:53 {
            errors
            health {
              lameduck 5s
            }
            ready
            kubernetes cluster.local in-addr.arpa ip6.arpa {
              pods insecure
              fallthrough in-addr.arpa ip6.arpa
              ttl 30
            }
            prometheus :9153
            forward . "/etc/resolv.conf"
            cache 30
            loop
            reload
            loadbalance
        }
```
- After deployment, verify coredns config map has the corefile as intended
- Edit cluster and upgrade the Kubernetes version
- Look at the CoreDNS config map and see it has returned to default

**Result**
Logs from upgrading indicates rke-user-addon job was not executed after upgrading

```
2021/10/14 20:34:35 [INFO] cluster [c-dnrn6] provisioning: [addons] Successfully saved ConfigMap for addon rke-user-addon to Kubernetes
2021/10/14 20:34:35 [INFO] cluster [c-dnrn6] provisioning: [addons] Executing deploy job rke-user-addon
2021/10/14 20:34:35 [DEBUG] [k8s] Job rke-user-addon-deploy-job already exists..
2021/10/14 20:34:35 [INFO] cluster [c-dnrn6] provisioning: [addons] User addons deployed successfully
```

But the CoreDNS job was replaced, which replaced the configmap with the default:

```
2021/10/14 20:34:23 [INFO] cluster [c-dnrn6] provisioning: [addons] Successfully saved ConfigMap for addon rke-coredns-addon to Kubernetes
2021/10/14 20:34:23 [INFO] cluster [c-dnrn6] provisioning: [addons] Executing deploy job rke-coredns-addon
2021/10/14 20:34:23 [DEBUG] [k8s] replacing job rke-coredns-addon-deploy-job..
2021/10/14 20:34:29 [DEBUG] [k8s] waiting for job rke-coredns-addon-deploy-job to complete..
```

**Expected Result**

`rke-user-addons` job is replaced similar to the rke-coredns-addon job, or the ability to configure a corefile to replace default that can be retained between upgrades. It seems reasonable to run the user addon job again on upgrade because the system job runs and changes what the user addon changed. The reason that it does not run in this case is because nothing has changed in the addon; Rancher believes it not necessary to re-run the job because it was already run successfully once.

**Screenshots**

**Additional context**

SURE-3508

**Comment by markusewalker** [comment]
**TEST PLAN**
1. Setup Rancher, create a standard user and login as that user.
2. Provision a downstream RKE1 node driver cluster to the second highest K8s version available.
     - While provisioning, edit the cluster's YAML and add the following underneath the `rancher_kubernetes_engine_config` block:
```
addons: |-
    ---
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: coredns
      namespace: kube-system
    data:
      Corefile: |
        .:53 {
            errors
            health {
              lameduck 5s
            }
            ready
            kubernetes cluster.local in-addr.arpa ip6.arpa {
              pods insecure
              fallthrough in-addr.arpa ip6.arpa
              ttl 30
            }
            prometheus :9153
            forward . "/etc/resolv.conf"
            cache 30
            loop
            reload
            loadbalance
        }
```
     - Open the browser's developer tools and view the traffic before pressing `Create`. Once done, create the cluster and validate in the `rancherKubernetesEngineConfig` block that the above is present.
3. Validate that the cluster and nodes appear in a status of `Active` in Cluster Management and Cluster Explorer.
4. Edit the cluster's YAML and validate that the custom ConfigMap has the corefile as specified.
5. Upgrade the cluster's K8s version while inspecting the network traffic in the browser's developer tools.
     - Validate once more in the `rancherKubernetesEngineConfig` block if the specified information is present.
6. Once upgraded, validate that the cluster and nodes remain in a state of `Active` in Cluster Management and Cluster Explorer.
7. Once more, edit the clusters YAML and validate if the custom ConfigMap still has the corefile specified.

**Comment by markusewalker** [comment]
Unable to reproduce this on `v2.6.2` and `v2.6-head` https://github.com/rancher/rancher/commit/b41b8b11b037d32b910403d9706cb8ab5ce48b0d.

**ENVIRONMENT DETAILS**
- Rancher install: Helm
- Rancher version:`v2.6.2` / `v2.6-head` https://github.com/rancher/rancher/commit/b41b8b11b037d32b910403d9706cb8ab5ce48b0d
- Browser: Chrome

**TEST RESULT**
UNABLE TO REPRODUCE

**VERIFICATION STEPS**
1. Setup Rancher, create a standard user and login as that user.
2. Provisioned a downstream RKE1 node driver cluster to the second highest K8s version available.
     - While provisioning, edit the cluster's YAML and add the following underneath the `rancher_kubernetes_engine_config` block:
```
addons: |-
    ---
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: coredns
      namespace: kube-system
    data:
      Corefile: |
        .:53 {
            errors
            health {
              lameduck 5s
            }
            ready
            kubernetes cluster.local in-addr.arpa ip6.arpa {
              pods insecure
              fallthrough in-addr.arpa ip6.arpa
              ttl 30
            }
            prometheus :9153
            forward . "/etc/resolv.conf"
            cache 30
            loop
            reload
            loadbalance
        }
```
     - Open the browser's developer tools and view the traffic before pressing `Create`. Once done, created the cluster and validatde in the `rancherKubernetesEngineConfig` block that the above is present.
3. Validate that the cluster and nodes appeared in a status of `Active` in Cluster Management and Cluster Explorer.
4. Edited the cluster's YAML and validated that the custom ConfigMap has the corefile as specified:
![image](https://user-images.githubusercontent.com/52591819/186773355-a80ce387-cdc7-454a-adce-98b2d5cd8a59.png) 
5. Navigated to Cluster Explorer > cluster > Workload > Storage > ConfigMaps.
6. Viewed the `coredns` YAML and validated that the corefile is preserved:
![image](https://user-images.githubusercontent.com/52591819/186773378-990b0348-8168-403a-9307-d722eb2ae7dd.png) 
7. Upgraded the cluster's K8s version while inspecting the network traffic in the browser's developer tools.
     - Validated once more in the `rancherKubernetesEngineConfig` block that the specified information is present.
8. Once upgraded, validated that the cluster and nodes remain in a state of `Active` in Cluster Management and Cluster Explorer.
9. Once more, edited the clusters YAML and validated that the custom ConfigMap still has the corefile specified:
![image](https://user-images.githubusercontent.com/52591819/186773421-d364a9e6-e8ec-44b2-ba3f-7c9f2ca7828d.png) 
10. Navigated back to Cluster Explorer > cluster > Workload > Storage > ConfigMaps.
11. Validated that the corefile is preserved in the `coredns`:
![image](https://user-images.githubusercontent.com/52591819/186773454-16cfa5c1-2113-41e6-8a1e-34c27441ff11.png)

**NOTES**
In `v2.6.2`, it is important to note that the `ttl 30` was removed and that STUBDOMAINS comment was added in the `coredns` ConfigMap. However, in the tested `v2.6-head` https://github.com/rancher/rancher/commit/b41b8b11b037d32b910403d9706cb8ab5ce48b0d, this is not the case and the ConfigMap appears exactly as originially it does in the cluster's YAML:
![image](https://user-images.githubusercontent.com/52591819/186773569-3f1238ff-e0ff-4d09-9e2b-bbfb5653c47b.png)

@jakefhyde I am unsure if this technically means that the RKE user addons was not retained, but at least from the above, I was unable to reproduce this issue. Perhaps we can sync offline so I can better understand where I am misunderstanding in my test plan.

**Comment by snasovich** [comment]
This seems to be the same as https://github.com/rancher/rancher/issues/36985 actually.

**Comment by snasovich** [comment]
@kinarashah , since you're working on a solution for https://github.com/rancher/rancher/issues/38749, this one seems to be related and it makes sense to account for both for the fix approach. Assigning to you.

**Comment by kinarashah** [comment]
Fix merged in rke `v1.4.0-rc1` and has been vendored into Rancher.

**Comment by markusewalker** [comment]
Validated this in `v2.7-head` https://github.com/rancher/rancher/commit/c567ef3f2bd13c2f2a5819242ab88cd334b1ae79.

**ENVIRONMENT DETAILS**
- Rancher install: Docker
- Rancher version:
     - Fresh install: `v2.7-head` https://github.com/rancher/rancher/commit/c567ef3f2bd13c2f2a5819242ab88cd334b1ae79
     - Rancher upgrade: `v2.6.8` -> `v2.7-head` https://github.com/rancher/rancher/commit/c567ef3f2bd13c2f2a5819242ab88cd334b1ae79
- Browser: Chrome

**TEST RESULT**
Scenario 1: Rancher Fresh Install - PASS
Scenario 2: Rancher Upgrade - PASS
Scenario 3: Standalone RKE - PASS

**VERIFICATION STEPS**
Scenario 1: Rancher Fresh Install
1. Setup Rancher, create a standard user and login as that user.
2. Provisioned a downstream RKE1 node driver cluster to the second highest K8s version available.
     - While provisioning, edit the cluster's YAML and add the following underneath the `rancher_kubernetes_engine_config` block:
```
addons: |-
    ---
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: coredns
      namespace: kube-system
    data:
      Corefile: |
        .:53 {
            errors
            health {
              lameduck 5s
            }
            ready
            kubernetes cluster.local in-addr.arpa ip6.arpa {
              pods insecure
              fallthrough in-addr.arpa ip6.arpa
              ttl 30
            }
            prometheus :9153
            forward . "/etc/resolv.conf"
            cache 30
            loop
            reload
            loadbalance
        }
```
3. Validated that the cluster and nodes appeared in a status of `Active` in Cluster Management and Cluster Explorer.
4. Edited the cluster's YAML and validated that the custom ConfigMap has the corefile as specified:
![image](https://user-images.githubusercontent.com/52591819/192619127-41b9b1dd-e36c-4573-a0a4-9cd505fc5add.png) 
5. Navigated to Cluster Explorer > cluster > Workload > Storage > ConfigMaps.
6. Viewed the `coredns` YAML and validated that the corefile is preserved.
7. Upgraded the cluster's K8s version.
8. Once upgraded, validated that the cluster and nodes remained in a state of `Active` in Cluster Management and Cluster Explorer and that the K8s version is upgraded.
9. Once more, edited the clusters YAML and validated that the custom ConfigMap still has the corefile specified:
![image](https://user-images.githubusercontent.com/52591819/192619186-a4e3bbfb-7594-4731-b294-3979d976cf69.png) 
10. Navigated back to Cluster Explorer > cluster > Workload > Storage > ConfigMaps.
11. Validated that the corefile is preserved in the `coredns`. 
_________

Scenario 2: Rancher Upgrade (`v2.6.8` -> `v2.7-head`)
1. Setup Rancher, create a standard user and login as that user.
2. Provisioned a downstream RKE1 node driver cluster to the second highest K8s version available.
     - While provisioning, edit the cluster's YAML and added the addons mentioned in Scenario 1.
3. Validated that the cluster and nodes appeared in a status of `Active` in Cluster Management and Cluster Explorer.
4. Edited the cluster's YAML and validated that the custom ConfigMap has the corefile as specified:
![image](https://user-images.githubusercontent.com/52591819/192619378-8f8c036f-2cda-443a-b3ed-ee263c8362dd.png) 
5. Navigated to Cluster Explorer > cluster > Workload > Storage > ConfigMaps.
6. Viewed the `coredns` YAML and validated that the corefile is preserved. 
7. Upgraded Rancher server to `v2.7-head` https://github.com/rancher/rancher/commit/c567ef3f2bd13c2f2a5819242ab88cd334b1ae79.
8. Validated that the cluster did not auto-upgrade the K8s version.
9. Edited the cluster's config and validated that the addons preserved.
10. Repeated steps 5-6 and ensured the corefile is also preserved.
11. Edited the cluster's config, upgraded the K8s version and upgraded the cluster.
12. Validated that the cluster and nodes are in an `Active` state and the K8s version is updated in Cluster Management and Cluster Explorer.
13. Repeated steps 10-11 and ensured everything has been preserved through the upgrade.

________

Scenario 3: Standalone RKE1
1. On a client machine, setup RKE version `v1.4.0-rc1`.
2. In the `cluster.yml`, added the following addon:

<img width="856" alt="image" src="https://user-images.githubusercontent.com/52591819/192619505-af3539b0-fda7-4292-8811-7076228c71a8.png"> 

3. Ran `rke up --config cluster.yml` and validated that the cluster came up successfully:
```
$ kubectl get nodes
NAME       STATUS   ROLES                      AGE     VERSION
mew-rke1   Ready    controlplane,etcd,worker   6m45s   v1.24.4
mew-rke2   Ready    controlplane,etcd,worker   6m45s   v1.24.4
mew-rke3   Ready    controlplane,etcd,worker   6m45s   v1.24.4
$ kubectl get pods -A
NAMESPACE       NAME                                       READY   STATUS      RESTARTS   AGE
ingress-nginx   ingress-nginx-admission-create-zrvjv       0/1     Completed   0          6m10s
ingress-nginx   ingress-nginx-admission-patch-zw7jg        0/1     Completed   1          6m10s
ingress-nginx   nginx-ingress-controller-28hbd             1/1     Running     0          6m10s
ingress-nginx   nginx-ingress-controller-57vzb             1/1     Running     0          6m10s
ingress-nginx   nginx-ingress-controller-dx76g             1/1     Running     0          6m10s
kube-system     calico-kube-controllers-74df54cbb7-chptw   1/1     Running     0          6m40s
kube-system     canal-c4h29                                2/2     Running     0          6m40s
kube-system     canal-gfwtw                                2/2     Running     0          6m40s
kube-system     canal-j8slk                                2/2     Running     0          6m40s
kube-system     coredns-59499769fb-hf778                   1/1     Running     0          6m29s
kube-system     coredns-59499769fb-twtqq                   1/1     Running     0          6m30s
kube-system     coredns-autoscaler-67cbd4599c-x8xpl        1/1     Running     0          6m30s
kube-system     metrics-server-585b7cc746-d2tdk            1/1     Running     0          6m20s
kube-system     rke-coredns-addon-deploy-job-8k9nv         0/1     Completed   0          6m32s
kube-system     rke-ingress-controller-deploy-job-p4r6n    0/1     Completed   0          6m12s
kube-system     rke-metrics-addon-deploy-job-fs7kv         0/1     Completed   0          6m22s
kube-system     rke-network-plugin-deploy-job-k6s6c        0/1     Completed   0          6m42s
kube-system     rke-user-addon-deploy-job-d6nfn            0/1     Completed   0          6m2s
```
4. Re-edited the cluster.yml and validated that the addons remained.
5. Re-ran `rke up --config cluster.yml` with no changes; validated that the `rke-user-addon` got recreated.
